haystack/preview/components/generators/openai/chatgpt.py

```diff
@@ -0,0 +1,275 @@
+        self.model_name = model_name
+        self.system_prompt = system_prompt
+
+        self.max_tokens = max_tokens
+        self.temperature = temperature
+        self.top_p = top_p
+        self.n = n
+        self.stop = stop or []
+        self.presence_penalty = presence_penalty
+        self.frequency_penalty = frequency_penalty
+        self.logit_bias = logit_bias or {}
+        self.stream = stream
+        self.streaming_callback = streaming_callback or default_streaming_callback
+        self.streaming_done_marker = streaming_done_marker
+
+        self.openai_organization = openai_organization
+        self.api_base_url = api_base_url
+
+        tokenizer = None
+        for model_prefix, tokenizer_name in OPENAI_TOKENIZERS.items():
+            if model_name.startswith(model_prefix):
+                tokenizer = tiktoken.get_encoding(tokenizer_name)
+                break
+        if not tokenizer:
+            raise ValueError(f"Tokenizer for model '{model_name}' not found.")
+        self.tokenizer = tokenizer
+
+        max_tokens_limit = None
+        for model_prefix, limit in OPENAI_TOKENIZERS_TOKEN_LIMITS.items():
+            if model_name.startswith(model_prefix):
+                max_tokens_limit = limit
+                break
+        if not max_tokens_limit:
+            raise ValueError(f"Max tokens limit for model '{model_name}' not found.")
+        self.max_tokens_limit = max_tokens_limit
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(
+            self,
+            api_key=self.api_key,
+            model_name=self.model_name,
+            system_prompt=self.system_prompt,
+            max_tokens=self.max_tokens,
+            temperature=self.temperature,
+            top_p=self.top_p,
+            n=self.n,
+            stop=self.stop,
+            presence_penalty=self.presence_penalty,
+            frequency_penalty=self.frequency_penalty,
+            logit_bias=self.logit_bias,
+            stream=self.stream,
+            # FIXME how to serialize the streaming callback?
+            streaming_done_marker=self.streaming_done_marker,
+            api_base_url=self.api_base_url,
+            openai_organization=self.openai_organization,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "ChatGPTGenerator":
+        """
+        Deserialize this component from a dictionary.
+        """
+        # FIXME how to deserialize the streaming callback?
+        return default_from_dict(cls, data)
+
+    @component.output_types(replies=List[List[str]])
+    def run(
+        self,
+        prompts: List[str],
+        api_key: Optional[str] = None,
+        model_name: Optional[str] = None,
+        system_prompt: Optional[str] = None,
+        max_tokens: Optional[int] = None,
+        temperature: Optional[float] = None,
+        top_p: Optional[float] = None,
```

Comment:
logprobs is missing, see https://platform.openai.com/docs/api-reference/completions/create#logprobs

Reply:
We're using the chat completion endpoint, which does not have logprobs (https://platform.openai.com/docs/api-reference/chat/create) : https://github.com/deepset-ai/haystack/pull/5692#discussion_r1311685583

However I realized we're missing `functions`, `function_call` and `user`, so for completeness I'll add those.